# Comparative genomics of DIR

How conserved are the proteins near iodate reductase (IdrA) across the phylogeny of IdrA and IdrA/AioA-like proteins?

## Dependencies

The following must be installed and in your path:
- [ncbi-genome-download](https://github.com/kblin/ncbi-genome-download)
- [MMSeqs](https://github.com/soedinglab/MMseqs2)
- [FastTree](http://www.microbesonline.org/fasttree/)
- [MUSCLE](http://www.drive5.com/muscle/muscle.html)
- [numpy](https://numpy.org/)
- [pandas](https://pandas.pydata.org/)
- [ete3](http://etetoolkit.org/)

## Methods

It is recommended to run this script as a python notebook. The script can be run as an stand along script, but there will be a fair amount of editing needed beforehand. Nonetheless, if you choose to run it as a script, be sure to make all the edits provided in the instructions before hand. 

### 1. Download genomes 
The first two cells of the notebook will allow you to download genomes to your directory. Make sure you use the directory structure in the repository.
Download genomes from a list of RefSeq/GenBank accessions. This list is named either:

- refseq-accessions.txt
- genbank-accessions.txt

Good practice here is to save old accession lists with the date as a prefix, and keep the above lists as your running active lists.
If you already downloaded genomes, toggle the download option to  False!
If you add new genomes delete old genome files and re-download.
```
download = False 
```
Have a duplicate problem?
Open a terminal window in the directory with you accessions list, and deduplicate your entries.  
```
$ sort genbank-accessions.txt | uniq -u > genbank-accessions_dedup.txt
``` 

## Define phylogeny

### 2. Identify HMM hit in each genome
The next cell will then search your dataset of downloaded genomes for hits matching your HMM. There are several considerations to take into account when creating an HMM, such as gene level indicators and meaningful cutoffs. For the sake of this tutorial, the HMM provided is for a combined AioA/IdrA HMM that was used to initially train an HMM that includes all AioA/IdrA proteins. A subsequent HMM was used to specifically distinguish IdrA and AioA using the gene level indicators produced by this script. 
For additional information on this, I recommend looking at the methods in our [publication](https://www.nature.com/articles/s41396-021-01034-5)

It is important to define the inputs here, as they will determine what you are searching for and how stringent your search is:
```
hmm = './data/hmm/combined_iriA_aioA.hmm'
threshold = 640
```

It is essential that you harmonize the dictionaries to your hits, otherwise the code breaks
```
for k in list(hmmhits.keys()):
    if k not in path_to_hits:
        del hmmhits[k]
with open('iodate_reducing_genomes.txt', 'w') as fh:
    for item in hmmhits: 
        fh.write(item + '\n')
```
This step will put out the total number of hits identified during the search and comapre it to those in the path. The number should be the same. 

### 3. Generate phylogenetic tree of HMM hits
This next step generates a phylogenetic tree as a newick file. 

Inputs here are important as well. 
- __faa_ingroup__: This are the protein sequences from your HMM hits placed into a single multifasta file
- __faa_outgroup__: Depending on your analysis, you are going to want to have sequences that you have verified as part of your outgroup. Since here we want to place the arsenite oxidase AioA as our outgroup, we use a multifasta containing known arsenite oxidases
- __temp__: This is a temporary file that the code generates while going through this cell. You should delete it before running the code if the file is present.  
- __faa__: This file is generated by the code and combines the ingroup and outgroup. Delete this file if present.
```
faa_ingroup = hits
faa_outgroup = './data/tree/aioA.faa' 
temp = './data/tree/temp.faa'
faa = './data/tree/iriA-all.faa'
```
A couple things to note here:
There is a filtering option for you to use:
```
filt = 80
f_aln = './data/tree/iriA-'+ str(filt) + '.aln'
filter_gaps = ' '.join(['python3', './scripts/remove-gapped-positions.py', '-p', str(filt), '-i', aln, '-o', f_aln])
sp.call(filter_gaps, shell=True)
```
The tree alignment doesn't use this option, but you can if you want. Just make sure to change:
```
fasttree = ' '.join(['fasttree','-boot 10000', aln, ">", tree])
```
to
```
fasttree = ' '.join(['fasttree','-boot 10000', f_aln, ">", tree])
```

Also note that the `-boot 10000` option is turned on. You can change this in case you want to change the number of bootstraps. I recommend looking at the fasttree documentation. 

After running this cell, you should have files with the extensions:
- .aln
- .nwk

The subsequent cell will do some pythonic magic to add human readable names to the accession numbers.

Finally, you will get to the point where you will draw your tree.
This will use the ete3 package and draw the tree inline. 

A few considerations at this point:
- Make sure to [set your outgroup](http://etetoolkit.org/docs/latest/tutorial/tutorial_trees.html#tree-rooting) so that you can meaningfully infer phylogeny on your tree
- Be sure to look at the ete3 documentation. There are a lot of things you can do with this package that allows you to color clades, and change tree styles.
- The support values are drawn onto the tree as black (>0.99), dark gray (>0.90), and light gray (>0.80). I don't recommend you change these, but you can. 

You will get two trees out of the cells that draw trees. Each has its utility. The expanded tree tells you the phylogeny of each hit. The collapsed tree shows you the different IdrA clades:

![image](https://user-images.githubusercontent.com/27031932/142703922-4c6e75fa-efa8-4ad4-be38-56bea9a9d248.png)

The default settings give you some good stuff, but I recommend changing it to your needs. 

### 4. Define gene neighborhood composition
Obtain genes within +/- 10 positions of HMM hits ("gene neigborhoods")
Group proteins from gene neighborhoods by sequence similarity ("subfamilies")

## Analysis
- Which subfamilies are conserved which clades?
- What the functions of those subfamilies?



